Adding basic review mechanism

EtiNoked opened the pull request:
Currently, the controllers render JSON structure, and there is a spec only for Restaurants controller

Review belongs to a restaurant and can be added via restaurant view (show). Also, the restaurant's rating is updated according to its reviews.

EtiNoked committed:
Adding basic review mechanism -
Review belongs to a restaurant, and can be added via restaurant view (show). Also, the restaurant's rating is updated according to its reviews.

For now, there is no option to edit\remove reviews, I would like to add Users to the application and then allow the user to update only his reviews.

EtiNoked committed:
Merge branch 'create_restaurants' of https://github.com/EtiNoked/WeEat into restaurant_reviews

Fixing conflict:
#	app/controllers/restaurants_controller.rb

EtiNoked committed:
Update Gemfile

EtiNoked committed:
Changing the css files to be compitable as the view, though it might be changed when I add the react components

EtiNoked committed:
Merge branch 'restaurant_reviews' of https://github.com/EtiNoked/WeEat into restaurant_reviews

EtiNoked committed:
Using before action in reviewsController as well

EtiNoked committed:
Merge branch 'master' of https://github.com/EtiNoked/WeEat into restaurant_reviews

Merging from master

pazaricha commented on app/controllers/restaurants_controller.rb:
```diff
@@ -5,9 +5,14 @@ class RestaurantsController < ApplicationController
 
   def index
     @restaurants = Restaurant.sorted_by_name_asc
+    @restaurants.each do |r|
```
This should not happen as part of the request but instead it should happen as a callback every time a review for a restaurant is created.

pazaricha commented on app/controllers/restaurants_controller.rb:
```diff
@@ -5,9 +5,14 @@ class RestaurantsController < ApplicationController
 
-  def show; end
+  def show
+    calculate_rating @restaurant
```
same as above, the controller should not handle the rating calculation

pazaricha commented on app/controllers/restaurants_controller.rb:
```diff
@@ -56,7 +61,13 @@ def set_cuisines
     @cuisines = Cuisine.sorted_by_name_asc
   end
 
+
```
minor - remove extra line space

pazaricha commented on app/controllers/reviews_controller.rb:
```diff
@@ -0,0 +1,59 @@
+  def edit; end
+
+  def update
+    if @review.update_attributes(review_params)
```
use update instead of update_attributes

EtiNoked committed:
1. Handle rating calculations inside the models instead of controllers.
2. Use 'update' instead of 'update_attributes'

tomerdesserww commented on app/assets/stylesheets/reviews.scss:
```diff
@@ -0,0 +1,3 @@
+// Place all the styles related to the review controller here.
```
please remove file

tomerdesserww commented on app/controllers/reviews_controller.rb:
```diff
@@ -0,0 +1,59 @@
+  before_action :find_review, :except => [:index, :new, :create]
+
+  def index
+    @restaurant_id = params.require(:restaurant_id)
```
Why assigning it as a class param? who else uses it?

tomerdesserww commented on app/controllers/reviews_controller.rb:
```diff
@@ -0,0 +1,59 @@
+
+  def index
+    @restaurant_id = params.require(:restaurant_id)
+    @reviews = Review.where(:restaurant_id => @restaurant_id)
```
please use : instead of =>, this is the style guide in WW
So Review.where(restaurant_id: @restaurant_id)

Also, please drop the assignment

tomerdesserww commented on app/controllers/reviews_controller.rb:
```diff
@@ -0,0 +1,59 @@
+class ReviewsController < ApplicationController
+
+  before_action :find_review, :except => [:index, :new, :create]
```
please change except to only (there are only 2 method where you use it

pazaricha commented on app/models/restaurant.rb:
```diff
@@ -4,7 +4,13 @@
   validates_presence_of :name
+
+  def update_rating
+    self.rating = reviews.sum(:rating) / reviews.length
```
just FYI this code will fail if someone will create a review and than delete it since it will do
`x / 0`

Myabe add:
```ruby
self.rating = if reviews.length > 0
                reviews.sum(:rating) / reviews.length
              else
                 0
              end
```

pazaricha replied:
This is still relevant.

EtiNoked committed:
Modify controllers to work with REST API

EtiNoked committed:
Clean! restaurant controller (using REST API)

EtiNoked committed:
Adding serializer to Restaurant

EtiNoked committed:
add .DS_Store & .idea to .gitignore, delete from cache

EtiNoked committed:
delete .DS_Store (in .gitignore)

EtiNoked committed:
add .DS_Store & .idea to .gitignore, delete from cache

EtiNoked committed:
Cleanup for restaurants_controller to render json structure & adding specs (at this point, just for restaurants controller)

EtiNoked committed:
Cleanup for controllers to render json structure

tomerdesserww commented on .idea/WeEat.iml:
```diff
@@ -29,6 +29,7 @@
     <orderEntry type="library" scope="PROVIDED" name="actionmailer (v5.2.1, rbenv: 2.3.1) [gem]" level="application" />
     <orderEntry type="library" scope="PROVIDED" name="actionpack (v5.2.1, rbenv: 2.3.1) [gem]" level="application" />
     <orderEntry type="library" scope="PROVIDED" name="actionview (v5.2.1, rbenv: 2.3.1) [gem]" level="application" />
+    <orderEntry type="library" scope="PROVIDED" name="active_model_serializers (v0.10.7, rbenv: 2.3.1) [gem]" level="application" />
```
please exclude this file

tomerdesserww commented on Gemfile:
```diff
@@ -29,7 +29,7 @@ gem 'jbuilder', '~> 2.5'
 # gem 'redis', '~> 4.0'
 # Use ActiveModel has_secure_password
-gem 'bcrypt', '~> 3.1.7'
+# gem 'bcrypt', '~> 3.1.7'
```
please clean up the gemfile from all redundant gems

tomerdesserww commented on app/controllers/reviews_controller.rb:
```diff
@@ -0,0 +1,40 @@
+  def create
+    render json: Review.create!(review_params), status: :created
+  rescue ActiveRecord::RecordInvalid => error
+    render json: {status: 'error', code: 400, message: error.message}, :status => 400
```
please convert all statuses to string symbols
400 --> bad_request

pazaricha replied:
better yet, remove all rescues. Rails gives this to you out of the box.

tomerdesserww commented on app/models/review.rb:
```diff
@@ -0,0 +1,13 @@
+class Review < ApplicationRecord
```
please use https://github.com/ctran/annotate_models to document the model's attributes

tomerdesserww commented on app/serializers/application_serializer.rb:
```diff
@@ -0,0 +1,3 @@
+class ApplicationSerializer < ActiveModel::Serializer
+  attributes :id
```
nice

tomerdesserww commented on app/serializers/restaurant_serializer.rb:
```diff
@@ -0,0 +1,4 @@
+# Restaurant serializer
+class RestaurantSerializer < ApplicationSerializer
+  attributes :id, :name, :cuisine_id, :rating, :ten_bis, :address
```
don't you get it from the application serilizer??

pazaricha replied:
exactly, you don't need the id here.

tomerdesserww commented on app/views/restaurants/show.html.erb:
```diff
@@ -30,4 +30,12 @@
     </tr>
   </table>
 
+  <br />
```
please delete all the erbs now, they're all redundant

tomerdesserww commented on app/views/reviews/_review_form.html.erb:
```diff
@@ -0,0 +1,16 @@
+<%= present_errors(f.object) %>
```
same

tomerdesserww commented on test/controllers/review_controller_test.rb:
```diff
@@ -0,0 +1,29 @@
+require 'test_helper'
```
please remove the tests

tomerdesserww commented on test/fixtures/reviews.yml:
```diff
@@ -0,0 +1,11 @@
+# Read about fixtures at http://api.rubyonrails.org/classes/ActiveRecord/FixtureSet.html
```
please remove

tomerdesserww commented on test/models/review_test.rb:
```diff
@@ -0,0 +1,7 @@
+require 'test_helper'
```
please remove

pazaricha commented on app/controllers/cuisines_controller.rb:
```diff
@@ -1,66 +1,38 @@
-    @cuisine = Cuisine.new
+  def show
+    render json: Cuisine.find(params[:id])
+  rescue ActiveRecord::RecordNotFound => error
```
remove this rescue and the code below it as rails does this for you automatically.

pazaricha commented on app/controllers/cuisines_controller.rb:
```diff
@@ -1,66 +1,38 @@
-      render('new')
-    end
+    render json: Cuisine.create!(cuisine_params), status: :created
+  rescue ActiveRecord::RecordInvalid => error
```
I believe Rails does this for you automatically as well. so you can remove it.

pazaricha commented on app/controllers/cuisines_controller.rb:
```diff
@@ -1,66 +1,38 @@
-      render('edit')
-    end
+    render json: Cuisine.find(params[:id]).update!(cuisine_params), status: :accepted
+  rescue ActiveRecord::RecordInvalid => error
```
remove both rescues

pazaricha commented on app/controllers/cuisines_controller.rb:
```diff
@@ -1,66 +1,38 @@
-      redirect_to(cuisines_path)
-    end
+    render json: Cuisine.find(params[:id]).destroy!, status: :accepted
+  rescue ActiveRecord::RecordNotFound => error
```
remove rescue

pazaricha commented on app/controllers/restaurants_controller.rb:
```diff
@@ -1,62 +1,39 @@
-    @restaurant = Restaurant.new({:name => 'Default'})
+  def show
+    render json: Restaurant.find(params[:id])
+  rescue ActiveRecord::RecordNotFound => error
```
remove all rescues from this file.

pazaricha commented on config/routes.rb:
```diff
@@ -16,5 +16,11 @@
 
+  resources :reviews do
+    member do
+      get :delete
```
if you remove all the `erb` files and just go with jsons you can remove this action.

pazaricha commented on spec/controllers/restaurants_controller_spec.rb:
```diff
@@ -0,0 +1,179 @@
+  let(:restaurant1) {create(:restaurant, name: 'R1')}
+  let(:restaurant2) {create(:restaurant, name: 'R2')}
+  let(:restaurant3) {create(:restaurant, name: 'R3')}
+  let(:restaurants) { [restaurant1, restaurant2, restaurant3]}
```
{ [restaurant1, restaurant2, restaurant3] }

pazaricha commented on spec/controllers/restaurants_controller_spec.rb:
```diff
@@ -0,0 +1,179 @@
+    create(:cuisine, name: 'C1')
+  end
+
+  let(:restaurant1) {create(:restaurant, name: 'R1')}
```
add 1 space after opening and before closing curly brackets. 
`{ create(:restaurant, name: 'R1') }`

pazaricha commented on spec/controllers/restaurants_controller_spec.rb:
```diff
@@ -0,0 +1,179 @@
+  let(:rest_params) {FactoryBot.attributes_for(:restaurant)}
+  let(:rest_no_name) {{cuisine_id: 1, delivery_time: 50}}
+  let(:rest_no_cuisine) {{name: 'Rest', delivery_time: 50}}
+
```
remove extra line spaces. make it just 1 line space.

pazaricha commented on spec/controllers/restaurants_controller_spec.rb:
```diff
@@ -0,0 +1,179 @@
+
+
+
+  describe 'Index' do
```
the convention it to not go with the capital letter on specs descriptions. I know it sounds weird but that's the common pattern.

EtiNoked committed:
Avoid the dividing by zero possibility by checking if reviews array is empty first.

pazaricha commented on spec/controllers/restaurants_controller_spec.rb:
```diff
@@ -0,0 +1,179 @@
+    context 'Returns all the restaurants sorted by name' do
+      before do
+        create_cuisine
+        allow(Restaurant).to receive(:sorted_by_name_asc).and_return(restaurants)
```
I would change it to `expect(Restaurant).to receive(:sorted_by_name_asc).and_return(restaurants)`

EtiNoked committed:
1. remove WeEat.iml from source control.
2. Remove CSS files
3. Use string symbol "bad_request" instead of its value (400)

pazaricha commented on spec/controllers/restaurants_controller_spec.rb:
```diff
@@ -0,0 +1,179 @@
+      end
+
+      it 'Checking the first restaurant in the array' do
+        expect(response_body.first['id']).to eq(restaurant1.id)
```
there is not much point in this test because you never actually call you `sorted_by_name_asc` method, you stub it so this will always be true. If you do want to test this `sorted_by_name_asc` scope, do it in the model and don't stub it.

EtiNoked committed:
Delete all erb files

EtiNoked committed:
Delete all tests files, since we are using specs instead

EtiNoked committed:
Remove handling exceptions code in the controllers since rails does it for me

EtiNoked committed:
Add model documentation using annotate

EtiNoked committed:
Remove id from restaurant serializer (redundant)

EtiNoked committed:
Adding serializer for reviews & improve restaurant's serializer a bit

EtiNoked committed:
Fix some comments in restaurant controller spec

EtiNoked committed:
remove delete from routes

GeReV commented on app/controllers/application_controller.rb:
```diff
@@ -1,6 +1,9 @@
 
 class ApplicationController < ActionController::Base
+
+  skip_before_action :verify_authenticity_token
```
This is due to JS-related errors, right?
I believe there is a possible fix along these lines (adding a header with a specific value): https://github.com/shakacode/react_on_rails/blob/c8bbb087c4760fdec86c0ccdd8f67045f93f0ce0/node_package/src/Authenticity.js

GeReV commented on spec/factories.rb:
```diff
@@ -0,0 +1,14 @@
+    name 'Rest_default'
+    cuisine_id 1
+    address 'Tel Aviv'
+    delivery_time 50
```
Why not use Faker and some random values?

GeReV reviewed (approved):
Looks great!

EtiNoked committed:
Adding faker to generate random data in the spec
